### PR TITLE
Handle nulls properly as arguments in the API

### PR DESCRIPTION
**Pull request by william-gr**

### CODE REVIEW

File: src/middlewared/middlewared/schema.py [hunk 1]
```diff
@@ -42,17 +45,28 @@ def clean(self, value):
...
         self.verbose = verbose or name
         self.validators = validators or []
         self.register = register
 
     def clean(self, value):
+        if value is None and self.null is False:
+            raise Error(self.name, 'null not allowed')
+        if value is NOT_PROVIDED:
+            if self.has_default:
+                return copy.copy(self.default)
```
Comment (themylogin) [inline src/middlewared/middlewared/schema.py]: Why not a deep copy?

File: src/middlewared/middlewared/schema.py [hunk 2]
```diff
@@ -118,9 +132,9 @@ class Str(EnumMixin, Attribute):
 
     def clean(self, value):
         value = super(Str, self).clean(value)
-        if value is None and not self.required:
-            return self.default
-        if isinstance(value, int):
+        if value is None:
+            return value
+        if isinstance(value, int) and not isinstance(value, bool):
```
Comment (themylogin) [inline src/middlewared/middlewared/schema.py]: I don't understand this check (` and not isinstance(value, bool)`) here and further, could you please explain?

Comment (william-gr) [reply src/middlewared/middlewared/schema.py]: @themylogin 

    >>> isinstance(False, int)
    True

Comment (themylogin) [reply src/middlewared/middlewared/schema.py]: Oh ****

File: src/middlewared/middlewared/schema.py [hunk 3]
```diff
@@ -433,6 +452,9 @@ def dump(self, value):
         return value
 
     def validate(self, value):
+        if value is None:
```
[the remark this answers is not in the document]
Comment (themylogin) [reply src/middlewared/middlewared/schema.py]: +1

**Comment by william-gr** [comment]
@themylogin @sonicaj I believe the reviews have been addressed, please check again.

**Pull request merged**